pkg/serviceaccount/metadata.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &issuerServer{
+		metadata: issuerMetadata{
+			Issuer:        iss,
+			JwksURI:       jwksURI,
+			ResponseTypes: []string{"id_token"}, // Kubernetes only produces ID tokens
+			SubjectTypes:  []string{"public"},   // https://openid.net/specs/openid-connect-core-1_0.html#SubjectIDTypes
+			SigningAlgs:   getAlgs(ks),          // REQUIRED by OIDC
+		},
+		keys: ks,
+	}, nil
+}
+
+// issuerMetadata provides a minimal subset of OIDC provider metadata:
+// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+type issuerMetadata struct {
+	Issuer string `json:"issuer"` // REQUIRED in OIDC; meaningful to relying parties.
+	// AuthzEndpoint string   `json:"authorization_endpoint"`                // REQUIRED in OIDC; but useless to relying parties.
+	JwksURI       string   `json:"jwks_uri"`                              // REQUIRED in OIDC; meaningful to relying parties.
+	ResponseTypes []string `json:"response_types_supported"`              // REQUIRED in OIDC
+	SubjectTypes  []string `json:"subject_types_supported"`               // REQUIRED in OIDC
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"` // REQUIRED in OIDC
+
+}
+
+// issuerServer is an HTTP server for metadata of the KSA token issuer.
+type issuerServer struct {
+	errorHandler func(error)
+	metadata     issuerMetadata
+	keys         *jose.JSONWebKeySet
+}
+
+// Install adds this server to the request router c.
+func (s *issuerServer) Install(c *restful.Container) {
+	// Container.Add "will detect duplicate root paths and exit in that case",
+	// so we need a root for /.well-known/openid-configuration to avoid conflicts.
+	cfg := new(restful.WebService)
+	cfg.Path(OpenIDConfigPath).Route(
+		cfg.GET("").
+			To(fromStandard(s.serveConfiguration)).
+			Doc("get serviceaccount issuer OIDC configuration").
+			Operation("getServiceAccountIssuerMetadata"))
+	c.Add(cfg)
+	// ...and another one for the JWKS
+	jwks := new(restful.WebService)
+	jwks.Path(JwksPath).Route(
+		jwks.GET("").
+			To(fromStandard(s.serveKeys)).
+			Doc("get serviceaccount issuer keys").
+			Operation("getServiceAccountIssuerKeys"))
+	c.Add(jwks)
+}
+
+// fromStandard provides compatibility between the standard (net/http) handler signature and the restful signature.
+func fromStandard(h http.HandlerFunc) restful.RouteFunction {
+	return func(req *restful.Request, resp *restful.Response) {
+		h(resp, req.Request)
+	}
+}
+
+func (s *issuerServer) serveConfiguration(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	// TODO(mtaufen): Set cache header appropriately
+
+	// TODO(mtaufen): trace the error handling here to understand how it works
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "\t")
+	// We can't stop writing the body and switch the header; but we can record when an error occurred.
+	err := enc.Encode(s.metadata)
+	if err != nil && s.errorHandler != nil {
```

Comment:
Why are we only handling errors if errorHandler is nil? What's the point of errorHandler?